src/Dscp/Educator/Txs/Type.hs

```diff
@@ -0,0 +1,62 @@
+
+-- | Definitions of private transactions
+
+module Dscp.Educator.Txs.Type
+       ( PrivateTx (..)
+       , PrivateTxWitness (..)
+       , PrivateTxAux (..)
+
+       -- * Lenses
+       , ptaTx
+       , ptaWitness
+       , ptGrade
+       , ptSignedSubmission
+       , ptTime
+       , ptwKey
+       , ptwSig
+       ) where
+
+import Control.Lens (makeLenses)
+import Data.Time.Clock (UTCTime)
+
+import Dscp.Core.Types (Grade, SignedSubmission (..))
+import Dscp.Crypto (Hash, PublicKey, Signature)
+
+-- | Private transaction.
+data PrivateTx = PrivateTx
+    { _ptSignedSubmission :: !SignedSubmission
+    -- ^ Every transaction contains one signed student submission
+    , _ptGrade            :: !Grade
+    -- ^ Grade for this submission
+    , _ptTime             :: !UTCTime
+    -- ^ Timestamp for this transaction
+    } deriving (Show, Eq, Generic)
+
+type PrivateTxId = Hash PrivateTx
+
+-- | Which data to sign in transaction.
+-- 'PrivateTxId' is basically a hash of all transaction contents,
+-- so it's sufficient to sign only that.
+type PrivateTxSigData = PrivateTxId
+
+-- | Type alias for private tx signature.
+type PrivateTxSig = Signature PrivateTxSigData
+
+-- | Witness contains data required to verify transaction.
+-- Included 'PublicKey' belongs either to Student or Educator.
+-- TODO: maybe we can say that Educator's key is already known
```

Comment:
Yes. That's important -- there is no point in storing educator signatures on the educator's side sometimes.

Reply:
Actually, this is some redundant stuff - we have student's signature in `SignedSubmission` datatype anyway. Types `PrivateTxWitness` and `PrivateTxAux` should be removed altogether (dunno how they're still here)